### ISSUE TITLE
[FEATURE] Toast Trait Exception Handler Support

### ISSUE THREAD

**Issue by NazimNasser** [robsontenorio/mary#150]
It would be really nice to have the ability to handle exceptions using the Toast trait in the exception handler

Right now using the Toast Trait in exception handler doesn't work because the js() function can't be used there.

**Comment by robsontenorio** [comment]
@NazimNasser  That would be a nice addition. Do you have any draft regarding this?

**Comment by NazimNasser** [comment]
@robsontenorio Sure!

Here's how an ExceptionHandler would look like:

```
<?php

namespace App\Exceptions;

use Throwable;
use Mary\Traits\Toast;
use Illuminate\Http\Exceptions\ThrottleRequestsException;
use Illuminate\Foundation\Exceptions\Handler as ExceptionHandler;

class Handler extends ExceptionHandler
{
    use Toast;
    /**
     * The list of the inputs that are never flashed to the session on validation exceptions.
     *
     * @var array<int, string>
     */
    protected $dontFlash = [
        'current_password',
        'password',
        'password_confirmation',
    ];

    /**
     * Register the exception handling callbacks for the application.
     */
    public function register(): void
    {
        $this->reportable(function (Throwable $e) {
            //
        });

        $this->renderable(function (ToastException $e) {
            $this->toast(
                type: $e->type,
                title: $e->title,
                description: $e->description,
                timeout: $e->timeout,
                position: $e->position,
                redirectTo: $e->redirectTo
            );
        });
        
        ....
```

Here's how the ToastException would look like:

```
<?php

namespace App\Exceptions;

use Exception;
use Illuminate\Support\Facades\Log;

class ToastException extends Exception
{
    /**
     * Create a new exception instance.
     *
     * @param string $type
     * @param string $title
     * @param string $description
     * @param string $position
     * @param int $timeout
     * @param string $redirectTo
     * @return void
     */
    public function __construct(
        public string $type,
        public string $title,
        public ?string $description = null,
        public string $position = 'toast-top toast-end',
        public int $timeout = 3000,
        public ?string $redirectTo = null
    ) {
        parent::__construct("ToastException: $title");

        $this->type = $type;
        $this->title = $title;
        $this->description = $description;
        $this->position = $position;
        $this->timeout = $timeout;
        $this->redirectTo = $redirectTo;
    }
}
```

and now usage would be to 

`throw new ToastException(.....)`

**Comment by robsontenorio** [comment]
@NazimNasser I will give a try and probably include in the next release. If you have other usage scenarios regarding this let me know.

**Comment by robsontenorio** [comment]
@NazimNasser  I was not able to get it working, because it get "trapped" on Livewire default  "inline modal" error.

Did you miss some piece of coding?

<img width="742" alt="image" src="https://github.com/robsontenorio/mary/assets/118955/09d138f9-6710-41f9-b6f8-7b186ec6c5bc">

**Comment by robsontenorio** [comment]
For future reference, here is one approach.

**Note this is not related to Mary itself, this is a Livewire stuff.** 
This is very tricky because Livewire does not play well with global error handler from Laravel. If you have a better approach, please, let me know.


 The main benefit of this approach is **to avoid try/catch on every component's method** to decide if shows a toast.

**TLDR:** 

- Create your own `AppException` for situations where you want to toast a generic error. 
- Override the `render()` method from `app/exceptions/Handler.php`.
- Some tweak on `resources/js/app.js`



**app/exceptions/AppException.php**
```php
<?php

namespace App\Exceptions;

use Exception;

class AppException extends Exception
{
    public function __construct(string $message = "", public ?string $description = null)
    {
        parent::__construct($message);
    }
}
```

**YOUR Livewire component**
```php

...

use Toast;

// THERE IS NO TRY/CATCH HERE !
public function save() {

  // If something goes wrong Payment will rise our custom `AppException`
  $payment = new Payment(...);
  $payment->pay();

   // Everything  ok, uses Mary's builtin toast
   $this->success('Done!')

}
```

**Payment.php**
```php
class Payment() {
  
   ...

   public function pay() {
       
      if($age < 25) {
          // This is our custom exception that will become a toast
          throw new AppException("You are too young!", "Ask for your mom.");  
      }

     // Everything is OK, continue here your stuff
  }

}
```



**app/exceptions/Handler.php**
```php

class Handler extends ExceptionHandler
{

    // Probably you don't wanna report this kind of exception.
    protected $dontReport = [
        AppException::class
    ];
     ....

    // Exceptions that returns JSONS pairs with `resources/js/app.js` logic
    public function render($request, Throwable $e)
    {
        $status = parent::render($request, $e)->getStatusCode();
        $isLivewire = $request->hasHeader('X-Livewire');

        // 419 error
        if ($status == 419) {
            return response()->json(['status' => 419], 500);
        }

        // If it is a custom app exception, uses Toast
        if ($e instanceof AppException && $isLivewire) {
            $toast = [
                'title' => $e->getMessage(),
                'description' => $e->description,
                'position' => 'toast-top toast-end',
                'icon' => Blade::render("<x-mary-icon class='w-7 h-7' name='o-x-circle' />"),
                'timeout' => '3000',
                'css' => 'alert-error'
            ];

            return response()->json(['toast' => $toast], 500);
        }

       // Otherwise , call default Laravel error render
      return parent::render($request, $e);
      
   }
}
```

**resources/js/app.js**
```js

// `preventDefault()` avoids  to show the Livewire's inline modal error

document.addEventListener('livewire:init', () => {
    Livewire.hook('request', ({fail}) => {
        fail(({status, content, preventDefault}) => {
            try {
                content = JSON.parse(content)

                if (status === 419 || content.status === 419) {
                    window.location.reload()
                    preventDefault()
                }

                if (content.toast) {
                    // This is already exposed by Mary on JS side.
                    toast({toast: content.toast});
                    preventDefault()
                }

               // More stuff if you want

            } catch (e) {
                console.log(e)
            }
        })
    })
})

```
